### PR TITLE
Change workspace init to keep hosts healthy on container errors

### LINKED ISSUE #3081: Don't set workspace hosts unhealthy if the error was due to the container
As @nicknytko pointed out, we currently set the workspace host to unhealthy if there is any error during launch: https://github.com/PrairieLearn/PrairieLearn/blob/05e00bfa577ec2d4edd0d40085ecb56419acc7c2/workspace_host/interface.js#L1124

However, these errors might be caused by problems with the workspace itself. Maybe the instructor set an invalid image name. Maybe the image has an error inside it that means it can't start up correctly.

Instead of always setting the host to unhealthy, we should just set the workspace to `stopped` (with an appropriate error message) and return. This would be probably be easier to implement if we rewrite this function as async/await rather than using `waterfall()`. It could be something like:

```
try {
    ...
    await _getInitialZipAsync(); // errors from this will be caught by the top-level catch which will set host to unhealthy
    ...
    try {
        await _pullImage(); // errors from this will be caught immediately
    } catch (err) {
        workspaceHelper.updateState(workspace_id, 'stopped', `Error! Unable to pull image. Click "Reboot" to try again. Detail: ${err}`);
        return; // immediately exit from the initSequence(); don't set host to unhealthy
    }
    ...
} catch (err) {
    workspaceHelper.updateState(workspace_id, 'stopped', `Error! Click "Reboot" to try again. Detail: ${err}`);
    markSelfUnhealthy((err2) => {...});
    ...
}
```

### PR DESCRIPTION
**Pull request by tdy**
Resolves #3081:

* [x] Rewrite `interface.initSequence()` as async/await rather than waterfall
* [x] Instead of always setting the host to unhealthy, just set workspace to `stopped` (with an appropriate error message) and return

### CODE REVIEW

File: workspace_host/interface.js [hunk 1]
```diff
@@ -1070,71 +1050,78 @@ function _startContainer(workspace, callback) {
         callback(null, workspace);
     });
 }
+const _startContainerAsync = util.promisify(_startContainer);
 
 // Called by the main server the first time a workspace is used by a user
-function initSequence(workspace_id, useInitialZip, res) {
+async function initSequenceAsync(workspace_id, useInitialZip, res) {
```
Comment (mwest1066) [inline workspace_host/interface.js:1056]: I love the new `async` version of this function! So much easier to read.

**Review by mwest1066 — APPROVED** [review]

**Comment by tdy** [comment]
* merged `master`
* also moved the hostname query inline: https://github.com/PrairieLearn/PrairieLearn/pull/3138/commits/d9f1e3cee7730c1e9dc8974639acea49bb98c7e9

**Comment by mwest1066** [comment]
Another conflict, I'm afraid.

**Comment by tdy** [comment]
> 
> 
> Another conflict, I'm afraid.

* conflict resolved
* also fixed a `launched_at` oversight from #3134 (could've sworn I tested it, so not sure how that slipped by)

**Comment by mwest1066** [comment]
> * added `logger.warn` when marking host unhealthy

Good idea.

**Pull request merged**